source/auth/auth.rst

```diff
@@ -1175,70 +1175,146 @@ If AWS authentication fails for any reason, the cache MUST be cleared.
-
-.. code:: idl
-
-  PrincipalStepRequest:
-    description: Driver’s opening request in saslStart
-    fields:
-      n:
-        description: "Name of the OIDC user Principal"
-        type: string
-        optional: true
-
-Note that the principal name is optional as it may be provided by the IDP in environments where only one IDP is used.  The username provided by the user MUST be used as the principalName.
-
-.. code:: idl
-
-  IdPServerInfo:
-    description: "The information used by callbacks to authenticate with the Identity Provider."
-    fields:
-      issuer:
-        description: >-
-            URL which describes the Authentication Server. This identifier should be
-            the iss of provided access tokens, and be viable for RFC8414
-            metadata discovery and RFC9207 identification.
-        type:string
-      clientId:
-        description: "Unique client ID for this OIDC client"
-        type: string
-      requestScopes:
-        description: "Additional scopes to request from IDP"
-        type: array<string>
-        optional: true
-
-Server will use principalName (n) if provided in the driver’s PrincipalStepRequest to select an appropriate IDP.  This IDP's configuration will be returned in the server’s response that will be used by the end-user to acquire an Access Token.
-
-This Access Token will be used as the JWT in the driver’s JwtStepRequest to complete authentication.
-
-.. code:: idl
-
-  JwtStepRequest:
-      description: "Client's request with signed token"
-      fields:
-          jwt:
-              description: "Compact serialized JWT with signature"
-              cpp_name: JWT
-              type: string
+As an example, given the OIDC token JWT string "abcd1234", the SASL conversation looks like:
+
+| C: :javascript:`{saslStart: 1, mechanism: "MONGODB-OIDC", payload: BinData(0, BSON({"jwt": "abcd1234"}))}`
+| S: :javascript:`{conversationId : 1, payload: BinData(0, "..."), done: true, ok: 1}`
+
+Where the sent ``payload`` is a generic binary blob containing a BSON-encoded document ``{"jwt": "abcd1234"}``.
+
+Access Token Caching
+````````````````````
+
+Drivers MUST cache the Access Token used to authenticate a connection on the
+connection object. Additionally, drivers MUST cache the most recent access Token
+on the ``MongoClient``. If any operation fails with ``ReauthenticationRequired``
+(391), the driver MUST clear the access Token cached on the connection object
+and and reauthenticate the connection as described in ``Reauthentication``
+section below.
+
+Speculative Authentication
+``````````````````````````
+
+Drivers MUST implement speculative authentication for MONGODB-OIDC during the
+``hello`` handshake. If there is a cached Access Token on the ``MongoClient``,
+use it for authentication. Otherwise, call the configured workload callback to
```

Comment:
I thought we were not calling any callbacks during speculative auth, since it could still be a lengthy process.

Reply:
My understanding is that the _intended_ use case of the custom machine-to-machine callback is to allow users to integrate with OIDC providers that the driver doesn't have built-in support for (using a machine-to-machine auth flow). In that case, it _should_ behave similarly to a built-in provider (e.g. AWS or Azure) and should be significantly faster than the "human" callback. However, we don't control the implementation, so a custom callback could take several seconds (it's subject to CSOT, so technically could take up to 10-30s), which is longer than we'd expect to wait for most other auth mechanisms.

Should the "no callbacks during speculative auth" rule be extended to the custom machine-to-machine callback? Is there a case where a slow callback is a problem for creating the speculative auth payload (i.e. the initial `hello` command) but not a problem for post-`hello` auth commands (e.g. `saslStart`)?

Reply:
I think the original problem was that we were worried about delays for non-authenticated hellos (ping), but we decided that this was not an issue. Later, we agreed that a server response might be missing a `speculativeAuthenticate`, and that this might be either because spec auth failed, or because it was removed by (for example) a proxy. From some of my Java notes:

> When the process is initiated during the initial handshake, speculative auth will be used. It is possible to know that speculative auth succeeded, but no reliable way to know it failed. Therefore, callbacks should not be invoked during the initial handshake.

At the time, it must have been more obvious to me why being unable to determine failure implied that (apparently all) callbacks were unworkable during initial auth (though using a cached token was ok - though, now I think we cannot tell the difference?), but now I do not immediately see why. But perhaps someone else can.

Reply:
If the initial auth succeeded, but `speculativeAuthenticate` was absent, then we would redundantly invoke the callback a second time (since the absence implies that it might have failed). We should not invoke callbacks spuriously, so callbacks should not be invoked during speculative auth.

Reply:
I believe the redundant callback invocation problem is resolved by the updated cache management strategy for the human auth flow. The new logic assumes a missing specauth response means nothing and continues with non-speculative authentication with the cached access token. If the non-speculative authentication fails with the cached access token, then call for a new access token, first using a refresh token, then getting new `IdpInfo` and calling the human callback.

I'll go back through the proposed algorithms and make sure they're consistent, but I think we can remove the requirement to skip the callback during speculative authentication.

Reply:
I simplified the speculative auth behavior to only apply to cached access tokens. The result is that no new access tokens are fetched by any means (callback, built-in, etc) during speculative auth.